storm-server/src/main/java/org/apache/storm/scheduler/resource/NormalizedResources.java

```diff
@@ -246,59 +259,133 @@ public boolean couldHoldIgnoringSharedMemory(NormalizedResources other) {
+                usedValue = 0.0;
+            } else {
+                usedValue = used.otherResources[i];
+            }
+            total += usedValue / totalValue;
+        }
+        //Adjust the divisor for the average to account for any skipped resources (those where the total was 0)
+        int divisor = 2 + otherResources.length - skippedResourceTypes;
+        if (divisor == 0) {
+            /*This is an arbitrary choice to make the result consistent with calculateMin.
+             Any value would be valid here, becase there are no (non-zero) resources in the total set of resources,
+             so we're trying to average 0 values.
+             */
+            return 100.0;
+        } else {
+            return (total * 100.0) / divisor;
         }
-        //To get the count we divide by we need to take the maximum length because we are doing an average.
-        return (total * 100.0) / (2 + Math.max(otherResources.length, used.otherResources.length));
     }
 
     /**
-     * Calculate the minimum resource usage percentage with this being the total resources and
-     * used being the amounts used.
+     * Calculate the minimum resource usage percentage with this being the total resources and used being the amounts used.
+     *
      * @param used the amount of resources used.
-     * @return the minimum percentage used 0.0 to 100.0.
+     * @return the minimum percentage used 0.0 to 100.0. Clamps to 100.0 in case there are no available resources in the total.
      */
     public double calculateMinPercentageUsedBy(NormalizedResources used) {
         double totalMemory = getTotalMemoryMb();
         double totalCpu = getTotalCpu();
+
+        if (LOG.isTraceEnabled()) {
+            LOG.trace("Calculating min percentage used by. Used CPU: {} Total CPU: {} Used Mem: {} Total Mem: {}"
+                + " Other Used: {} Other Total: {}", totalCpu, used.getTotalCpu(), totalMemory, used.getTotalMemoryMb(),
+                toNormalizedOtherResources(), used.toNormalizedOtherResources());
+        }
+
+        if (used.otherResources.length > otherResources.length) {
+            throwBecauseResourceIsMissingFromTotal(used.otherResources.length);
+        }
+
         if (used.otherResources.length != otherResources.length
             || totalMemory == 0.0
             || totalCpu == 0.0) {
             //If the lengths don't match one of the resources will be 0, which means we would calculate the percentage to be 0.0
             // and so the min would be 0.0 (assuming that we can never go negative on a resource being used.
             return 0.0;
         }
-        double min = used.getTotalMemoryMb() / totalMemory;
-        min = Math.min(min, used.getTotalCpu() / getTotalCpu());
+
+        double min = 100.0;
+        if (totalMemory != 0.0) {
+            min = Math.min(min, used.getTotalMemoryMb() / totalMemory);
+        }
+        if (totalCpu != 0.0) {
```

Comment:
Again the math looks OK, but could you explain that 0 % used is ignored in the javadocs, because it is not obvious to me that it what will happen.